### PR TITLE
added rate limit

### PR DESCRIPTION
**Pull request by shivam-ef**
Rate limit is expressed in percentage of total vault tokens that can be minted in a specific time period. Hundred percent is 1 gwei.

### CODE REVIEW

File: src/LrtSquare.sol [hunk 1]
```diff
@@ -356,6 +426,23 @@ contract LrtSquare is
...
+
+        for (uint256 i = 0; i < len; ) {
+            uint256 valueOfTokenInVault = getAvsTokenTotalValuesInEth(tokens[i]);
+            percentagesInVault[i] = SafeCast.toUint64((valueOfTokenInVault * HUNDRED_PERCENT_LIMIT) / vaultTotalValue);
+            unchecked {
+                ++i;
+            }
+        }
+
+        return percentagesInVault;
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:442]: let's return the list of addresses `tokens` together so that this API can be more useful

File: src/LrtSquare.sol [hunk 2]
```diff
@@ -157,9 +204,18 @@ contract LrtSquare is
             1 * 10 ** decimals()
         );
 
+        uint256 totalSupplyBeforeDeposit = totalSupply();
         uint256 shareToMint = previewDeposit(_tokens, _amounts);
+        // check rate limit
+        if (totalSupplyBeforeDeposit !=0) {
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:210]: rate limit on deposit amount

File: src/LrtSquare.sol [hunk 3]
```diff
@@ -157,9 +204,18 @@ contract LrtSquare is
...
+            uint64 amountForLimit = SafeCast.toUint64((shareToMint * HUNDRED_PERCENT_LIMIT) / totalSupplyBeforeDeposit);
+            if(!rateLimit.consume(amountForLimit)) revert RateLimitExceeded();
+        } 
+        
         _deposit(_tokens, _amounts, shareToMint, _receiver);
 
+        _checkPercentagesInVault();
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:217]: limit on the max holdings per token

File: src/LrtSquare.sol [hunk 4]
```diff
@@ -340,6 +396,20 @@ contract LrtSquare is
         return total_eth;
     }
 
+    function getAvsTokenTotalValuesInEth(
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:399]: `getTotalValueLockedInEthOf` ?

File: src/LrtSquare.sol [hunk 5]
```diff
@@ -413,6 +501,23 @@ contract LrtSquare is
         return IERC20Metadata(erc20).decimals();
     }
 
+    function _checkPercentagesInVault() internal view {
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:504]: lets make it more descriptive. 

- `_checkPercentagesInVault` -> `_verifyPositionLimits`
- `maxPercentageInVault` -> `positionLimitPercentage`

File: src/LrtSquare.sol [hunk 6]
```diff
@@ -413,6 +501,23 @@ contract LrtSquare is
...
+        uint256 len = tokens.length;
+        uint64[] memory percentagesInVault = new uint64[](len);
+        uint256 vaultTotalValue = getVaultTokenValuesInEth(totalSupply());
+
+        if(vaultTotalValue == 0) return;
+
+        for (uint256 i = 0; i < len; ) {
+            uint256 valueOfTokenInVault = getAvsTokenTotalValuesInEth(tokens[i]);
+            percentagesInVault[i] = SafeCast.toUint64((valueOfTokenInVault * HUNDRED_PERCENT_LIMIT) / vaultTotalValue);
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:513]: Having a public view function which returns the position weight would be useful

`getPositionWeight(address token)`?

File: src/LrtSquare.sol [hunk 7]
```diff
@@ -27,18 +29,23 @@ contract LrtSquare is
...
     UUPSUpgradeable
 {
+    using BucketLimiter for BucketLimiter.Limit;
     using SafeERC20 for IERC20;
     using Math for uint256;
 
     struct TokenInfo {
         bool registered;
         bool whitelisted;
+        uint64 maxPercentageInVault;
```
Comment (seongyun-ko) [inline src/LrtSquare.sol:39]: `positionWeightLimit`?

File: src/LrtSquare.sol (hunk 1 above)
Comment (Shivam78288) [reply src/LrtSquare.sol:442]: Adding that

File: src/LrtSquare.sol (hunk 7 above)
Comment (Shivam78288) [reply src/LrtSquare.sol:39]: Done

File: src/LrtSquare.sol (hunk 6 above)
Comment (Shivam78288) [reply src/LrtSquare.sol:513]: Done

File: src/LrtSquare.sol (hunk 5 above)
Comment (Shivam78288) [reply src/LrtSquare.sol:504]: Done

**Pull request closed without merging**